Include some extra plugin samples

OdatNurd opened the pull request:
This includes some new example plugins to the plugins folder, showing
off some more common questions and also providing some items that are
generally useful to people at large.

 * A command to clear the console
 * Inserting text up to a specific column (e.g. whitespace)
 * Key binding context example for knowing if the cursor is at the EOF
 * Expanding snippets with multiple versions of selected text at once
 * Changing the local font size of a tab
 * Scrolling the viewport without moving the cursor
 * Scrolling the viewport to put the selected text at the top of the tab
 * Navigate a file by scopes
 * Navigate a file by arbitrary regex
 * Update a header in a file whenever a file is saved
 * Change the cursor when there is more than one selection

OdatNurd committed:
Include some extra plugin samples

This includes some new example plugins to the plugins folder, showing
off some more common questions and also providing some items that are
generally useful to people at large.

 * A command to clear the console
 * Inserting text up to a specific column (e.g. whitespace)
 * Key binding context example for knowing if the cursor is at the EOF
 * Expanding snippets with multiple versions of selected text at once
 * Changing the local font size of a tab
 * Scrolling the viewport without moving the cursor
 * Scrolling the viewport to put the selected text at the top of the tab
 * Navigate a file by scopes
 * Navigate a file by arbitrary regex
 * Update a header in a file whenever a file is saved
 * Change the cursor when there is more than one selection

keith-hall commented on plugins/README.md:
```diff
@@ -131,10 +151,42 @@ of each file for more information on full usage.
+
+ * [selection_to_top.py](selection_to_top.py) is a quick little command I made
+   for making it easier to illustrate code samples in my YouTube videos; it
+   scrolls the current file so that the cursor is the first line in the view.
```
```suggestion
   scrolls the current file so that the cursor is the first line in the
   viewport.
```

keith-hall commented on plugins/eof_context.py:
```diff
@@ -0,0 +1,49 @@
+        # NOTE: the whole selection must appear inside the last line for this
+        # to match; a selection that is only partially in the last line is
+        # not contained in the last line.
+        sels = [last_line.contains(s) for s in view.sel()]
```
it could make sense to use a generator here, to short circuit the comparison and save time if there are lots of selections...
```suggestion
        sels = (last_line.contains(s) for s in view.sel())
```

OdatNurd replied:
That's a good point; fixed.

keith-hall commented on plugins/insert_to_column.py:
```diff
@@ -0,0 +1,51 @@
+#
+# This implements an insert_to_column command that will insert whitespace at
+# every cursor that exists to move the cursor to the column specified. When
+# this is 0, the next closest ruler is used, with colunn 72 being a fallback
```
```suggestion
# this is 0, the next closest ruler is used, with column 72 being a fallback
```

keith-hall commented on plugins/insert_to_column.py:
```diff
@@ -0,0 +1,51 @@
+            col = col or self.find_next_ruler(caret, rulers)
+            spaces = col - caret
+            if spaces > 0:
+                self.view.insert(edit, sel.a, spaces * " ")
```
interesting, I somehow didn't know that "number * string" works as well as "string * number" :)

OdatNurd replied:
😁

keith-hall commented on plugins/pattern_navigate.py:
```diff
@@ -0,0 +1,55 @@
+        # Look in the given direction for the first match from the current
+        # position; if one is found jump there.
+        pick = lambda p: (point < p.a) if forward else (point > p.a)
+        for pos in regions if forward else reversed(regions):
```
does it make sense to bisect here to avoid looping through a potentially large number of selections?

OdatNurd replied:
That seems like a good idea; I need to work out the logic of that for the link/anchor navigation in HyperHelp as well. I'm going to table that for the time being and revisit this in another PR so we can get this one cleared away, if that's cool with you.

keith-hall replied:
Yep :)

keith-hall commented on plugins/scope_navigate.py:
```diff
@@ -0,0 +1,51 @@
+        # Look in the given direction for the first match from the current
+        # position; if one is found jump there.
+        pick = lambda p: (point < p.a) if forward else (point > p.a)
+        for pos in regions if forward else reversed(regions):
```
nice code re-use here :)

btw this scrap reminds me of one of mine:

```py
import sublime
import sublime_plugin


class SelectByScopeSelectorCommand(sublime_plugin.TextCommand):
    def run(self, edit, scope_selector=None, last_scope_only=True, auto_select=False):
        if not scope_selector:
            scope_at_first_caret = self.view.scope_name(self.view.sel()[0].a)
            if last_scope_only:
                scope_at_first_caret = scope_at_first_caret.split()[-1]

            if auto_select:
                scope_selector = scope_at_first_caret
            else:
                self.view.window().show_input_panel('Scope Selector', scope_at_first_caret, lambda value: self.view.run_command('select_by_scope_selector', { 'scope_selector': value }), None, None)
                return

        regions = self.view.find_by_selector(scope_selector)
        if regions:
            self.view.sel().clear()
            self.view.sel().add_all(regions)
            self.view.show_at_center(self.view.sel()[0])
        else:
            self.view.window().status_message('Unable to find anything matching selector "' + scope_selector + '"')
```

one thing I like about mine is - you can default the scope selector to the scope under the caret. And prompt the user to enter it. And that I refer to "scope selectors" instead of scopes ;)
Anyway, just food for thought - I like what you've done here and how it works in a similar way to another scrap.

OdatNurd replied:
That's interesting! Is that in the scraps repo? A quick search didn't find it, but things are going a little weird this evening in general...

keith-hall replied:
Nope, private atm / just a stackoverflow answer I believe. I also need to get round to posting stuff on this repo!

keith-hall commented on plugins/selectionless_scroll_lines.py:
```diff
@@ -0,0 +1,24 @@
+# This implements a selectionless_scroll_lines command that is a drop in
+# replacement for the scroll_lines command that is in the Sublime core; it
+# does the same thing but scrolls the view without changing the location of
+# the cursors as the built in command does.
```
```suggestion
# the cursors as the built-in command does.
```

keith-hall commented on plugins/selectionless_scroll_lines.py:
```diff
@@ -0,0 +1,24 @@
+# does the same thing but scrolls the view without changing the location of
+# the cursors as the built in command does.
+#
+# The following key bindings show how to use it (they overwrite the built in
```
```suggestion
# The following key bindings show how to use it (they overwrite the built-in
```

keith-hall commented on plugins/selectionless_scroll_lines.py:
```diff
@@ -0,0 +1,24 @@
+# the cursors as the built in command does.
+#
+# The following key bindings show how to use it (they overwrite the built in
+# key bindings on the built in command, blocking them from being used via the
```
```suggestion
# key bindings on the built-in command, blocking them from being used via the
```

keith-hall commented on plugins/snippets_with_selections.py:
```diff
@@ -0,0 +1,25 @@
+# Related Reading:
+#     https://forum.sublimetext.com/t/parse-two-selected-lines-in-html-tags/56162
+
+# This plugin implementd a version of the insert_snippet command that provides
```
```suggestion
# This plugin implements a version of the insert_snippet command that provides
```

keith-hall commented on plugins/snippets_with_selections.py:
```diff
@@ -0,0 +1,25 @@
+# This plugin implementd a version of the insert_snippet command that provides
+# access to each individual selection as numbered variables, such as
+# SELECTION_1, SELECTION_2, and so on. This allows an expansion to include
+# multuple selections in the replacement.
```
```suggestion
# multiple selections in the replacement.
```

keith-hall commented on plugins/update_last_edited.py:
```diff
@@ -0,0 +1,47 @@
+    as formatted by the given date format.
+    """
+    def run(self, edit, hdr="Last Edited: ", fmt="%d %b %Y %I:%M%p"):
+        span = self.view.find(f'{hdr}.*$', 0)
```
it may make sense to regex escape the `hdr`. Also, curious why not call it `header`? :)

OdatNurd replied:
The short name was entirely for brevity reasons so that it would fit in a code block in the forum without scrolling. I'm cool with changing it to a longer name generally though.

OdatNurd replied:
Not sure I get why this would benefit from a regex escape; doesn't that only matter for constant strings and not ones generated at runtime? I have expanded out the names of the arguments to be longer and more readable though.

keith-hall replied:
I'm thinking that if the command is run with a user-supplied `header` argument as opposed to using the default value defined here, it could consist of anything - parens, or other characters which have special meaning in regex, and if we want to treat it as a literal, it would need to be escaped? To be clear, its not about escaping for the f-string, but the regex engine - otherwise after the f-string is resolved, if the header was `*** last modification: `, the regex pattern could be e.g. `*** last modification: .*$` and would have different meaning than if we escaped the asterisks: `\*\*\* last modification: .*$`

Am I making sense? :)

OdatNurd replied:
Total sense! I've added in some quoting action on the header string as it comes in. It seems to be a bit overzealous, such as quoting spaces which seems unnecessary, but since I didn't notice this as an issue clearly I'm a boob anyway.

keith-hall commented on plugins/visible_multi_selection.py:
```diff
@@ -0,0 +1,45 @@
+#     https://forum.sublimetext.com/t/difference-when-multiple-selection/58087
+#
+# This simple plugin makes it easier to visualize when there is more than one
+# cursor active in the current file (particuarly when one or more of them may
```
```suggestion
# cursor active in the current file (particularly when one or more of them may
```

keith-hall reviewed (commented):
nice work! I have a few suggestions - mostly just small typos :)

OdatNurd committed:
Fix some typos and clean code up a bit

This includes some fixes related to silly typos and phrasing and also
cleans up some code smells as found by @keith-hall.

There are still a couple of outstanding items from the review to
revisit; a question is still outstanding regarding the use of a regex
quote on an input string, and the recommendation to use bisect() is a
good one but we're going to kick that one to a different PR.

OdatNurd commented:
I've addressed the spelling issues and some of the code related items. I'm not entirely sure what the regex suggestion is since I think that only applies to static strings and not those dynamically created at runtime?

I also like the idea of the bisect, but the logic of that is escaping me at the moment and also I need to come up with something similar for HyperHelp's link navigation, so I'd like to kick that to a separate PR to get this one closed out and everything updated. Thoughts?

OdatNurd committed:
Merge branch 'master' into new-plugins

OdatNurd committed:
Quote the incoming header string for safety

As recommended by @keith-hall, this escapes all of the regex special
characters in the incoming header string, since it's used as a part of
a regex as part of the command and we don't want it to go all sideways
on us.
